src/Bicep.Core/TypeSystem/DeployTimeConstantDirectViolationVisitor.cs

```diff
@@ -0,0 +1,127 @@
+
+        public override void VisitResourceAccessSyntax(ResourceAccessSyntax syntax)
+        {
+            this.FlagIfAccessingEntireResourceOrModule(syntax);
+        }
+
+        public override void VisitFunctionCallSyntax(FunctionCallSyntax syntax)
+        {
+            this.FlagIfFunctionRequiresInlining(syntax);
+
+            base.VisitFunctionCallSyntax(syntax);
+        }
+
+        public override void VisitInstanceFunctionCallSyntax(InstanceFunctionCallSyntax syntax)
+        {
+            this.FlagIfFunctionRequiresInlining(syntax);
+
+            base.VisitInstanceFunctionCallSyntax(syntax);
+        }
+
+        private void FlagIfAccessingEntireResourceOrModule(SyntaxBase syntax)
+        {
+            if (this.DeployTimeConstantContainer is not IfConditionSyntax and not ForSyntax)
+            {
+                // We can skip validation if we are inside a resource/module body or a function call, because the
+                // type checker should be able to produce type errors if the reference is invalid. There are also
+                // locations where referencing an entire resource/module module body is expected (e.g., scope).
+                return;
+            }
+
+            switch (this.SemanticModel.Binder.GetParent(syntax))
+            {
+                // var foo = [for x in [...]: {
+                //   bar: myVM <-- accessing an entire resource/module.
+                // }]
+                case not PropertyAccessSyntax and not ArrayAccessSyntax when
+                    this.TryExtractResourceOrModuleSymbolAndBodyType(syntax, false) is ({ } accessedSymbol, { } accessedBodyType):
+                    {
+                        this.FlagDeployTimeConstantViolation(syntax, accessedSymbol, accessedBodyType);
+
+                        return;
+                    }
+                // var foo = [for x in [...]: {
+                //   bar: myVNets[1] <-- accessing an entire resource/module via an array index.
+                // }]
+                case ArrayAccessSyntax { IndexExpression: IntegerLiteralSyntax } arrayAccessSyntax when
+                    this.SemanticModel.Binder.GetParent(arrayAccessSyntax) is not PropertyAccessSyntax and not ArrayAccessSyntax &&
+                    this.TryExtractResourceOrModuleSymbolAndBodyType(syntax, true) is ({ } accessedSymbol, { } accessedBodyType):
+                    {
+                        this.FlagDeployTimeConstantViolation(syntax, accessedSymbol, accessedBodyType);
+
+                        return;
+                    }
+
+                default:
+                    return;
+            }
+        }
+
+        private void FlagIfPropertyNotReadableAtDeployTime(SyntaxBase syntax, string propertyName, DeclaredSymbol accessedSymbol, ObjectType accessedBodyType)
+        {
+            if (accessedBodyType.Properties.TryGetValue(propertyName, out var propertyType) &&
+                !propertyType.Flags.HasFlag(TypePropertyFlags.ReadableAtDeployTime))
```

Comment:
I don't think it's necessary here unless this is being called a **lot**, but @rynowak made an interesting observation [here](https://github.com/Azure/bicep/blob/4ffccbccb22694501de24b752b567dbea1a14190/src/Bicep.Core/Parsing/Parser.cs#L153) about perf of `HasFlag` vs bitwise operations.

Reply:
That's good to know. I was searching for details about this and I happened to find a [benchmark](https://www.code4it.dev/blog/hasflag-performance-benchmarkdotnet) testing `HasFlag` vs bitwise operations, which suggests that we don't need to do the optimization to eliminate boxing/unboxing by ourselves. The .NET team has implemented some JIT optimization to generate bitwise operations for us since .NET core 2.1, which is documented in https://devblogs.microsoft.com/dotnet/ryujit-just-in-time-compiler-optimization-enhancements/.